---
title: "1-prepare_data.Rmd"
author: "Antton Alberdi"
date: "2023-12-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries

```{r libraries}
library(tidyverse)
library(ape)
```

## Retrieve and load data

### Count table
This is the document containing the number of sequencing reads from each sample have been mapped to each MAG. Note that this is the raw data that needs to be further processed before running any statistics on them.

```{r load_count}
counts <- read_tsv("https://sid.erda.dk/share_redirect/gR0yYR6oKL/results/stats/coverm_genome_2023-11-03_swine_mannan_mg.count.tsv") %>%
  rename_all(~ str_remove_all(., " Read Count")) #simplify column names
```

### Base hit table
This is the document containing the number of nucleotide bases have been covered by at least one read in each sample and MAG. This information is used to calculate MAG coverage values.

```{r load_hits}
basehits <- read_tsv("https://sid.erda.dk/share_redirect/gR0yYR6oKL/results/stats/coverm_genome_2023-11-03_swine_mannan_mg.covered_bases.tsv") %>%
  rename_all(~ str_remove_all(., " Covered Bases")) #simplify column names
```

### MAG lengths
This information is needed to calculate coverage values and to normalise read counts by genome length.
```{r load_maglengths}
maglengths <- read_tsv("https://sid.erda.dk/share_redirect/gR0yYR6oKL/results/stats/coverm_genome_2023-11-03_swine_mannan_mg.length.tsv") %>%
  rename(MAGlength=2) %>% #remove redundancy and rename
  select(Genome,MAGlength)
```

### MAG tree
This is the raw tree generated by GTDBtk, which needs to be pruned to obtain the phylogenetic tree of the MAGs.
```{r load_tree}
archaeatree <- read.tree(url("https://sid.erda.dk/share_redirect/BLJUHsGfWu/results/dereplicate/gtdbtk/classify/gtdbtk.ar53.classify.tree"))
bacteriatree <- read.tree(url("https://sid.erda.dk/share_redirect/BLJUHsGfWu/results/dereplicate/gtdbtk/classify/gtdbtk.backbone.bac120.classify.tree"))
magtree <- bind.tree(archaeatree, bacteriatree)
magtree$tip.label <- str_replace_all(magtree$tip.label,"'", "") #remove single quotes in MAG names
magtree <- keep.tip(magtree, tip=maglengths$Genome) # keep only MAG tips
```

### MAG taxonomy
This is the raw taxonomy table generated by GTDBtk, which is simplified for downstream analyses.
```{r load_tree}
taxonomy <- read_tsv("https://sid.erda.dk/share_redirect/BLJUHsGfWu/results/dereplicate/gtdbtk/gtdbtk.summary.tsv") %>%
  rename(Genome = user_genome) %>%
  mutate(Genome = str_replace_all(Genome,"\\.fa", "")) %>%
  separate(classification, c("Domain", "Phylum","Class","Order","Family","Genus","Species"),  sep =";") %>%
  select(Genome,Domain,Phylum,Class,Order,Family,Genus,Species)
```

### MAG functional annotations

This is the raw annotation table generated by DRAM, which is used to generate GIFT data using distillR.
```{r load_annotations}
annotations <- read_tsv("https://sid.erda.dk/share_redirect/BLJUHsGfWu/results/dereplicate/dram/annotations.tsv") %>%
  rename(gene=1,genome=2)
```

## Transform and normalise data

### Generate coverage table
By dividing the number of base hits by the length of each MAG, coverage values can be calculated.

```{r calc_coverage}
coverage <- basehits %>%
  mutate(across(where(is.numeric), ~ ./maglengths$MAGlength))
```
