---
title: "3-composition_overview.Rmd"
author: "Antton Alberdi"
date: "2023-12-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load libraries

```{r libraries, warning=FALSE, comments="", message=FALSE}
suppressPackageStartupMessages(library(tidyverse))
suppressPackageStartupMessages(library(ape))
suppressPackageStartupMessages(library(distillR))
suppressPackageStartupMessages(library(ggtree))
suppressPackageStartupMessages(library(ggnewscale))
suppressPackageStartupMessages(library(ggtreeExtra))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(phytools))
suppressPackageStartupMessages(library(Rtsne))
suppressPackageStartupMessages(library(ggh4x))
suppressPackageStartupMessages(library(hilldiv2))
```

## Load data

```{r load_data}
load("data/data.Rdata")
```

## Sort samples

```{r sort_samples}
#Arranged by days
samples_days <- sample_metadata %>%
  arrange(day,treatment,animal)

#Arranged by animal
samples_animal <- sample_metadata %>%
  arrange(treatment, animal, day)
```


## Taxonomy barplot (digesta)

```{r barplot_digesta, fig.height=8, fig.width=10, fig.fullwidth=TRUE}

#Get phylum colors from the EHI standard
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
  right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    select(colors) %>%
    pull()

barplot_digesta <- genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  filter(type=="digesta")  %>% #retain only digesta samples
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
    facet_grid(.~treatment,  scales="free_x") + #facet days
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")

barplot_digesta
```

## Taxonomy barplot (faeces)

```{r barplot_faeces_days, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
barplot_faeces_days <- genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  filter(type=="faeces") %>% #retain only faecal samples
  mutate(sample = factor(sample, levels = unique(samples_days$sample))) %>% #sort per animal code
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
    facet_nested(. ~ day + treatment,  scales="free_x") + #facet per day and treatment
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")

barplot_faeces_days
```

```{r barplot_faeces_animals, fig.height=8, fig.width=10, fig.fullwidth=TRUE}
barplot_faeces_animals <- genome_counts %>%
  mutate_at(vars(-genome),~./sum(.)) %>% #apply TSS nornalisation
  pivot_longer(-genome, names_to = "sample", values_to = "count") %>% #reduce to minimum number of columns
  left_join(., genome_metadata, by = join_by(genome == genome)) %>% #append genome metadata
  left_join(., sample_metadata, by = join_by(sample == sample)) %>% #append sample metadata
  filter(type=="faeces") %>% #retain only faecal samples
  mutate(sample = factor(sample, levels = unique(samples_days$sample))) %>% #sort samples per sampling day
  mutate(animal = factor(animal, levels = unique(samples_animal$animal))) %>% #sort animals per treatment
  ggplot(., aes(x=sample,y=count, fill=phylum, group=phylum)) + #grouping enables keeping the same sorting of taxonomic units
    geom_bar(stat="identity", colour="white", linewidth=0.1) + #plot stacked bars with white borders
    scale_fill_manual(values=phylum_colors) +
    labs(y = "Relative abundance") +
    facet_nested(. ~ treatment + animal,  scales="free_x") + #facet per treatment and animal
    guides(fill = guide_legend(ncol = 1)) +
    theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1),
          axis.title.x = element_blank(),
          panel.background = element_blank(),
          panel.border = element_blank(),
          panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.line = element_line(linewidth = 0.5, linetype = "solid", colour = "black")) +
   labs(fill="Phylum")

barplot_faeces_animals
```


## Alpha diversities
```{r alpha_div, message=FALSE, warning=FALSE, fig.height=10, fig.width=10, fig.fullwidth=TRUE}

#Calculate Hill numbers
richness <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=0) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(richness=1) %>%
            rownames_to_column(var="sample")

neutral <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(neutral=1) %>%
            rownames_to_column(var="sample")

phylogenetic <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,tree=genome_tree) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(phylogenetic=1) %>%
            rownames_to_column(var="sample")

# Aggregate basal GIFT into elements
dist <- genome_gifts %>%
    to.elements(., GIFT_db) %>%
    traits2dist(., method="gower")

functional <- genome_counts %>% 
            column_to_rownames(var="genome") %>% 
            select(where(~!all(. == 0))) %>% 
            hilldiv(.,q=1,dist=dist) %>% 
            t() %>% 
            as.data.frame() %>%
            rename(functional=1) %>%
            rownames_to_column(var="sample") %>%
            mutate(functional = if_else(is.nan(functional), 1, functional))



#Merge into a single table
#alpha_div <- cbind(sample=colnames(genome_counts[-1]),richness=q0n,neutral=round(q1n,3),phylo=round(q1p,3),func=round(q1f,3)) %>%
alpha_div <- richness %>%
      full_join(neutral,by=join_by(sample==sample)) %>%
      full_join(phylogenetic,by=join_by(sample==sample)) %>%
      full_join(functional,by=join_by(sample==sample)) %>%
      pivot_longer(-sample, names_to = "data", values_to = "value") %>%
      left_join(., sample_metadata, by = join_by(sample == sample)) %>%
      filter(library != "control") %>%
      mutate(value = as.numeric(value)) %>%
      mutate(type = factor(type, levels = c("faeces","digesta"))) %>% #sort sample types
      mutate(data = factor(data, levels = c("richness","neutral","phylogenetic","functional"))) #sort metrics


#Plot diversities
alpha_plot <- alpha_div %>%
        ggplot(aes(x=value, y=sample)) +
            geom_bar(stat='identity', fill="#6c9ebc") +
            facet_nested(type + day ~ data,  scales="free") + #facet per treatment and animal
            coord_cartesian(xlim = c(1, NA)) +
            theme_classic() +
            theme(
                strip.background = element_blank(),
                panel.grid.minor.x = element_line( size=.1, color="grey" ),
                axis.title.x = element_blank(),
                axis.title.y = element_blank(),
                axis.text.x = element_text(angle = 45, hjust = 1)
            )

alpha_plot
```
